### ISSUE TITLE
qemu://session?

### ISSUE THREAD

**Issue by lostinblue** [digitalocean/go-libvirt#51]
I don't want to run VMs as root, one would think naturally. When I specify the user socket I get:

````
2017/12/01 18:53:19 failed to connect: internal error: unexpected QEMU URI path '/system', try qemu:///session
````

So I searched through the code for system but I could not find anything that would be obvious that I could change to session. Any guidance on this topic would be appreciated for both this library and the qemu one you maintain. If this a repeat I apologize, but I did look through the issues of both libraries and did not see anything that was obviously relevant.

**Comment by benlemasurier** [comment]
Hi @lostinblue, this may help: https://libvirt.org/uri.html#URI_qemu

What is the full connection URI you are using?

**Comment by lostinblue** [comment]
I have been studying that page and have not had luck unfortunately. 

I have tried qemu for the network and session for the address but that had failure. So I tried using the socket, by doing unix network and using /run/user/1000/libvirt/libvirt-sock as the address which generated this error. I tried also libvirtd as the network which has the same failure as qemu as the network. So I'm not sure if the URI page is helpful unfortunately. Thanks for the quick response.

**Comment by lostinblue** [comment]
Oh I got it working with the libvirt/libvirt-go package but I would rather use a pure go implementation. I'm guessing its somehow hardcoded to use the root system QEMU path which seems like an odd choice because of potential VM breakouts leading to full root access of the hypervisor.

**Comment by benlemasurier** [comment]
Can you give me the exact connection URI you're using, even better a sample piece of code that's failing?

**Comment by lostinblue** [comment]
Thanks for taking time to help me out with this issue, I would really like to use and contribute to this library. 

So I'm using the very basic example code:

````
package main

import (
	"fmt"
	"log"
	"net"
	"time"

	"github.com/digitalocean/go-libvirt"
)

func main() {
	//c, err := net.DialTimeout("tcp", "127.0.0.1:16509", 2*time.Second)
	//c, err := net.DialTimeout("tcp", "192.168.1.12:16509", 2*time.Second)
	c, err := net.DialTimeout("unix", "session", 2*time.Second)
	if err != nil {
		log.Fatalf("failed to dial libvirt: %v", err)
	}

	l := libvirt.New(c)
	if err := l.Connect(); err != nil {
		log.Fatalf("failed to connect: %v", err)
	}

	v, err := l.Version()
	if err != nil {
		log.Fatalf("failed to retrieve libvirt version: %v", err)
	}
	fmt.Println("Version:", v)

	domains, err := l.Domains()
	if err != nil {
		log.Fatalf("failed to retrieve domains: %v", err)
	}

	fmt.Println("ID\tName\t\tUUID")
	fmt.Printf("--------------------------------------------------------\n")
	for _, d := range domains {
		fmt.Printf("%d\t%s\t%x\n", d.ID, d.Name, d.UUID)
	}

	if err := l.Disconnect(); err != nil {
		log.Fatal("failed to disconnect: %v", err)
	}
}
````

**Example 1**
c, err := net.DialTimeout("unix", "/var/run/libvirt/libvirt-sock", 2*time.Second)
*2017/12/02 22:59:43 failed to connect: authentication required*

**Example 2**
c, err := net.DialTimeout("unix", "/run/user/1000/libvirt/libvirt-sock", 2*time.Second)
*2017/12/02 23:01:35 failed to connect: internal error: unexpected QEMU URI path '/system', try qemu:///session*

So I tried both unix sockets. I tried every variation of using qemu://session I could think of.

**Comment by benlemasurier** [comment]
Dialing "qmp", or "qemu" will never work. `unix`, `tcp`, `tcp6`, etc.. are the [only valid network identifiers](https://golang.org/pkg/net/#Dial). As for the socket or tcp authentication methods, you'll want to take a look at your libvirtd configuration, specifically under the sections "UNIX socket access controls" and "Authentication".

**Comment by YorikSar** [comment]
@lostinblue You are right, it is hardcoded to use `qemu:///system` URL here: https://github.com/digitalocean/go-libvirt/blob/master/rpc.go#L123 - you can try to change it to `qemu:///session` to check if that's causing your problem.

**Comment by galaxyblack** [comment]
Would you accept a pull request where I make this a configurable option? We need session not system

**Comment by trapgate** [comment]
Yes, it would be good to address this.
